### PR TITLE
ceph-bluestore-tool list-objects operation

### PR DESCRIPTION
**Pull request by aclamk**
Adding list-objects operation to ceph-bluestore-tool. 
It list in detail content of Onode "O" prefixes in RocksDB.
It allows to inspect in detail state of object internal metadata.

Options for filtering objects and reducing output tree are also available "--select", "--trim".

### CODE REVIEW

**Comment by aclamk** [comment]
jenkins make api tests

**Comment by aclamk** [comment]
jenkins test api

**Comment by aclamk** [comment]
jenkins retest this please

File: src/common/formatter_filter.cc [hunk 1]
```diff
@@ -0,0 +1,1583 @@
...
+#include "common/safe_io.h"
+
+#include "os/bluestore/BlueFS.h"
+#include "os/bluestore/BlueStore.h"
+#include "common/admin_socket.h"
+#include "common/url_escape.h"
+
+#include "formatter_filter.h"
+/*
+  expression:= literal ||
```
Comment (tchaikov) [inline src/common/formatter_filter.cc:26]: this handwriting tokenizer and parser are difficult to understand. i'd suggest rewrite it with `boost::spirit`. i can help with this part if needed.

File: src/common/formatter_filter.cc [hunk 2]
```diff
@@ -0,0 +1,1583 @@
...
+	  default:
+	    token_literal.push_back(peeked.front());
+	    peeked = peeked.substr(1);
+	    break;
+	  }
+	}
+	token = eError;
+	break;
+      }
+    case '0' ... '9':
```
Comment (tchaikov) [inline src/common/formatter_filter.cc:122]: ahh, just realized that case range is also supported by Clang.

File: src/common/formatter_filter.cc (hunk 1 above)
Comment (aclamk) [reply src/common/formatter_filter.cc:26]: I like the idea.
If examples in crush parsing will not be enough, I will capitalize on your offer of help.

File: src/os/bluestore/bluestore_dump.cc [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+
+#include "os/bluestore/BlueFS.h"
+#include "os/bluestore/BlueStore.h"
+#include "common/admin_socket.h"
+#include "common/url_escape.h"
+namespace po = boost::program_options;
+
+#include "bluestore_dump.h"
+
+void Dump::dump(const bluestore_blob_t& t)
```
Comment (ifed01) [inline src/os/bluestore/bluestore_dump.cc:27]: What's I don't like about this and below dump() func implementations is that they are set apart from corresponding  structs. May be define such dumps as members of relevant structs/classses?

File: src/os/bluestore/BlueStore.cc [hunk 4]
```diff
@@ -414,6 +414,23 @@ static int get_key_object(const S& key, ghobject_t *oid)
   return 0;
 }
 
+int BlueStore::objkey_to_oid(const std::string& key, ghobject_t *oid)
+{
+  return get_key_object(key, oid);
+}
+
+int BlueStore::extentkey_to_oid(const std::string& key, ghobject_t *oid, uint32_t *offset)
```
Comment (ifed01) [inline src/os/bluestore/BlueStore.cc:422]: using 'extent' here and below is confusing - this object is extent shard - a bunch of extents of an onode which use separate KV entry for the sake of having smaller entries to operate on. Hence suggest either use shard or extent_shard or something

File: src/os/bluestore/bluestore_dump.cc [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+      buffer = bf->attach(f);
+      f = ff->attach(buffer, [&](){matched = true;});
+    }
+    std::string key = obj_it->key();
+    ghobject_t oid;
+    bool is_oid = *key.rbegin() == 'o';
+    if (is_oid)
+      dump_object(key, obj_it->value());
+    else
+      dump_extent(key, obj_it->value());
```
Comment (ifed01) [inline src/os/bluestore/bluestore_dump.cc:427]: IMO it's better not to dump extent shards separately (they are just physical parts of a single logical entity - Onode) Instead onode should load everything it needs and dump itself as a whole logical entity.

File: src/os/bluestore/bluestore_dump.cc [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+	f->dump_int("bytes", t.bytes_per_au[i]);
+      }
+      f->close_section();
+    }
+  }
+}
+
+
+//equivalent to BlueStore::ExtentMap::decode_spanning_blobs
+int Dump::decode_spanning_blobs(std::map<int16_t, bluestore_blob_t>& spanning_blobs,
```
Comment (ifed01) [inline src/os/bluestore/bluestore_dump.cc:106]: Can we use all such equivalents from BlueStore directly instead? Well, most of such copy-pastes will go away anyway if dump implementations are members of respective classes.

File: doc/dev/bluestore.rst [hunk 7]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
 | 100 byte clone overwrite | P | N  | P | N        | P | N       | P | N        | N (F?)        |
 +--------------------------+--------+--------------+-------------+--------------+---------------+
+
+
+
+==========
+ Database
+==========
+
+BlueStore uses key-value database to store its state.
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:93]: uses a key-value database

File: doc/dev/bluestore.rst [hunk 8]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
+Currently only RocksDB has been integrated.
+
+
+Objects
+-------
+
+Objects and shards are stored in DB under prefix 'O'.
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:100]: in the DB with the prefix

File: doc/dev/bluestore.rst [hunk 9]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
+
+- Key is a serialization of *ghobject_t* oid.
+- Value is a serialization of TODO
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:103]: Is `TODO` the intended text here, or is it a placeholder that wasn't circled back to?

File: doc/dev/bluestore.rst [hunk 10]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
++------------+------+------------+------------------------------+
+| snap       | 8    | uint64_t   | oid.hobj.snap.val            |
++------------+------+------------+------------------------------+
+| generation | 8    | uint64_t   | oid.generation               |
++------------+------+------------+------------------------------+
+| suffix     | 1    | 'o'        | equals ONODE_KEY_SUFFIX      |
++------------+------+------------+------------------------------+
+
+*Lexistring* encodes binary string into readable string in a manner that preserves lexicographic order:
+
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:134]: encodes a binary string as an alphanumeric string in a manner

File: doc/dev/bluestore.rst [hunk 11]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
+a. Values <= '#' (0x23) encoded as '#'(hexdigit)(hexdigit).
+b. Values > '~' (0x7e) encoded as '~'(hexdigit)(hexdigit).
+c. Values '$'..'~' are not changed.
+d. Value '!' terminates string.
+
+In BS code function ``get_key_object()`` and ``get_object_key()`` are responsible to translate ghobject_t to serialized form.
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:140]: Suggestion, if I understand this correctly:

Within BlueStore code, the functions
are used to encode ``ghobject_t`` in serialized format.

File: doc/dev/bluestore.rst [hunk 12]
```diff
@@ -83,3 +83,60 @@ might be several blobs to consider (e.g., we might be able to *W* into one or
...
+
+**Note** Serialization of _hash_ entry breaks lexicographic order of objects by name in database.
```
Comment (anthonyeleven) [inline doc/dev/bluestore.rst:142]: in the database

**Review by anthonyeleven — APPROVED** [review]

**Pull request closed without merging**